opm/input/eclipse/EclipseState/Util/OrderedMap.hpp

```diff
@@ -20,14 +20,67 @@
 #ifndef OPM_ORDERED_MAP_HPP
 #define OPM_ORDERED_MAP_HPP
 
+#include <fmt/format.h>
+
 #include <unordered_map>
 #include <vector>
 #include <string>
 #include <stdexcept>
 #include <iterator>
+#include <string>
+#include <sstream>
+#include <set>
+#include <iostream>
+#include <cctype>
+#include <algorithm>
 
 namespace Opm {
 
+namespace OrderedMapDetail
+{
+template<class T, class A>
+//typename std::enable_if_t<!std::is_same<T,std::string>::value, std::string>
+std::string
+findSimilarStrings(const std::string&,
+                   const std::vector<T,A>&)
+{
+    return {};
+}
+
+template<class A, class K>
+std::string
+findSimilarStrings(std::string str,
+                   const std::vector<std::pair<std::string, K>,A>& storage)
+{
+    auto toUpper = [](const char c){ return std::toupper(c);};
+    std::transform(str.begin(), str.end(), str.begin(), toUpper);
+    std::set<std::string> alternatives;
+
+    for(const auto& entry: storage)
+    {
+        std::string upper = entry.first;
+        std::transform(upper.begin(), upper.end(), upper.begin(),
+                       toUpper);
+
+        if(upper.find(str) != std::string::npos || str.find(upper) != std::string::npos)
+        {
+            alternatives.insert(entry.first);
+        }
+    }
+
+    if (alternatives.empty())
+    {
+        return {};
+    }
+
+    std::stringstream concated;
+    std::copy(alternatives.begin(), alternatives.end(),
+              std::ostream_iterator<std::string>(concated, ", "));
+    auto concatedStr = concated.str();
+    return concatedStr.substr(0, concatedStr.size()-2);
```

Comment:
Since you're already using it, you should know that Libfmt happens to provide the very useful helper function [`fmt::join()`](https://fmt.dev/latest/api.html) which makes this kind of manipulation very easy.  As an example, the programme
```C++
#include <iostream>
#include <string>
#include <vector>

#include <fmt/format.h>

int main()
{
    const auto x = std::vector<int>{ 1, 2, 3, 4 };
    const auto y = fmt::format("{}", fmt::join(x, ", "));
    std::cout << "y = [ " << y << " ]\n";
}
```
will print
```
y = [ 1, 2, 3, 4 ]
```
to the console.

Reply:
Ouch, this made me realize that we cannot use libfmt in headers! Need to revisit this.

Reply:
> this made me realize that we cannot use libfmt in headers! Need to revisit this.

I don't really understand why not since we're guaranteed to have at least a bundled copy of the library, but if you're convinced that we can't use libfmt in headers then you should also look into removing the offending include statements from

* [`opm/simulators/wells/BlackoilWellModel_impl.hpp`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/opm/simulators/wells/BlackoilWellModel_impl.hpp#L33)
* [`opm/simulators/wells/GasLiftGroupInfo.hpp`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/opm/simulators/wells/GasLiftGroupInfo.hpp#L42)
* [`opm/simulators/wells/GasLiftSingleWell.hpp`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/opm/simulators/wells/GasLiftSingleWell.hpp#L33)
* [`opm/simulators/wells/GasLiftCommon.hpp`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/opm/simulators/wells/GasLiftCommon.hpp#L27)
* [`opm/simulators/wells/StandardWell.hpp`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/opm/simulators/wells/StandardWell.hpp#L53)
* [`opm/simulators/flow/FlowMainEbos.hpp`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/opm/simulators/flow/FlowMainEbos.hpp#L39)
* [`ebos/equil/initstateequil.hh`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/ebos/equil/initstateequil.hh#L52)
* [`ebos/equil/equilibrationhelpers.hh`](https://github.com/OPM/opm-simulators/blob/3a6abf7f5c2f43e291ab1b97ac2e9c39b31ffbfd/ebos/equil/equilibrationhelpers.hh#L32)
* [`opm/input/eclipse/Deck/FileDeck.hpp`](https://github.com/OPM/opm-common/blob/4d76351111bb90306d82d4c6dada95f1fbf627bc/opm/input/eclipse/Deck/FileDeck.hpp#L29)

and possibly others which I did not identify through a quick search of the source code.